### ISSUE TITLE
dhcp nics should put their hostname in /etc/hostname.if

### ISSUE THREAD

**Issue by richlowe** [joyent/smartos-live#595]
`dhcpagent` will request the hostname specified in `/etc/hostname.<if>` from the DHCP server.  At present, unfortunately, the `mdata-get` service blows away `/etc/hostname.<if>`, for dhcp NICS, and doesn't create one.  It appears this makes it impossible to request a hostname at all.

The service should probably do something like

```
echo "inet $(mdata-get sdc:hostname)" > /etc/hostname.${iface}
```

in the dhcp case

**Comment by richlowe** [comment]
A workaround patching the `/lib/svc/method/mdata-fetch` to do what I described above:

```diff
--- /lib/svc/method/mdata-fetch Thu Mar 31 03:07:02 2016
+++ local-hacks/mdata-fetch     Fri Apr  1 19:41:29 2016
@@ -230,7 +230,8 @@
     [[ -z ${ip} ]] && continue;
 
     if [[ ${ip} == "dhcp" ]]; then
-      touch /etc/dhcp.${iface}
+        touch /etc/dhcp.${iface}
+        echo "inet $(mdata-get sdc:hostname)" > /etc/hostname.${iface}
     elif [[ ${ip} == "addrconf" ]]; then
       touch /etc/addrconf.${iface}
     elif isIPv4 ${ip} && [[ -n ${netmask} ]]; then
```

Worked in the trivial case I cared about.  I'm not sure if there would be downsides in other deployments (presumably, at the least, you'd want to check a hostname was actually set in the metadata).

**Comment by melloc** [comment]
@richlowe I've filed [OS-5301](https://smartos.org/bugview/OS-5301) for this. In addition to mdata-fetch, the net-physical script will also need to be updated.

**Comment by richlowe** [comment]
@melloc it seemed like the patch above did enough (as does a service that does basically the same thing, but saves me from having to carry a locally modified mdata-fetch).

What's missing out of `network/physical`?

**Comment by melloc** [comment]
@richlowe `network/physical` checks to see if that file exists and tries to process it. It will need to skip lines starting with `inet` so that setups using both static addresses and DHCP and have a hostname will be able to work.

**Comment by melloc** [comment]
Resolved in 26d614b.
